File: test/ch21-benchmarks/bench-004-binary-tree.jl
# BENCH-004: Binary tree allocation/deallocation - Julia
# Tests: memory allocator, Julia GC, pointer chasing

mutable struct TreeNode
    left::Union{TreeNode, Nothing}
    right::Union{TreeNode, Nothing}
end

TreeNode() = TreeNode(nothing, nothing)

function make_tree(depth::Int64)::TreeNode
    if depth <= 0
        return TreeNode()
    end
    return TreeNode(make_tree(depth - 1), make_tree(depth - 1))
end

function check_tree(node::TreeNode)::Int64
    if node.left === nothing
        return 1
    end
    return 1 + check_tree(node.left) + check_tree(node.right)
end

function main()
    max_depth = 16
    min_depth = 4

    # Stretch tree
    stretch_depth = max_depth + 1
    stretch_tree = make_tree(stretch_depth)
    stretch_check = check_tree(stretch_tree)

    # Long-lived tree
    long_lived_tree = make_tree(max_depth)

    # Create and destroy many trees
    total_checks = 0
    for depth in min_depth:2:max_depth
        iterations = 1 << (max_depth - depth + min_depth)
        for _ in 1:iterations
            tree = make_tree(depth)
            total_checks += check_tree(tree)
        end
    end

    # Final checksum
    long_check = check_tree(long_lived_tree)

    # Silent for benchmarking
end

main()

File: test/ch21-benchmarks/bench-004-binary-tree.c
// BENCH-004: Binary tree allocation/deallocation - C
// Tests: memory allocator, pointer chasing, manual memory management

#include <stdio.h>
#include <stdlib.h>

typedef struct TreeNode {
    struct TreeNode *left;
    struct TreeNode *right;
} TreeNode;

TreeNode* make_tree(int depth) {
    TreeNode *node = (TreeNode*)malloc(sizeof(TreeNode));
    if (!node) {
        return NULL;
    }

    if (depth <= 0) {
        node->left = NULL;
        node->right = NULL;
    } else {
        node->left = make_tree(depth - 1);
        node->right = make_tree(depth - 1);
    }

    return node;
}

int check_tree(TreeNode *node) {
    if (node->left == NULL) {
        return 1;
    }
    return 1 + check_tree(node->left) + check_tree(node->right);
}

void free_tree(TreeNode *node) {
    if (node == NULL) {
        return;
    }
    free_tree(node->left);
    free_tree(node->right);
    free(node);
}

int main() {
    int max_depth = 16;
    int min_depth = 4;

    // Stretch tree
    int stretch_depth = max_depth + 1;
    TreeNode *stretch_tree = make_tree(stretch_depth);
    int stretch_check = check_tree(stretch_tree);
    free_tree(stretch_tree);

    // Long-lived tree
    TreeNode *long_lived_tree = make_tree(max_depth);

    // Create and destroy many trees
    int total_checks = 0;
    for (int depth = min_depth; depth <= max_depth; depth += 2) {
        int iterations = 1 << (max_depth - depth + min_depth);
        for (int i = 0; i < iterations; i++) {
            TreeNode *tree = make_tree(depth);
            total_checks += check_tree(tree);
            free_tree(tree);
        }
    }

    // Final checksum
    int long_check = check_tree(long_lived_tree);
    free_tree(long_lived_tree);

    // Silent for benchmarking
    return 0;
}
